drivers/sensor/pmw3360/pmw3360.c

```diff
@@ -0,0 +1,938 @@
+static int pmw3360_async_init_fw_load_start(struct pmw3360_data *dev_data)
+{
+	int err = 0;
+
+	/* Read from registers 0x02-0x06 regardless of the motion pin state. */
+	for (u8_t reg = 0x02; (reg <= 0x06) && !err; reg++) {
+		u8_t buf[1];
+		err = reg_read(dev_data, reg, buf);
+	}
+
+	if (err) {
+		LOG_ERR("Cannot read from data registers");
+		return err;
+	}
+
+	/* Write 0 to Rest_En bit of Config2 register to disable Rest mode. */
+	err = reg_write(dev_data, PMW3360_REG_CONFIG2, 0x00);
+	if (err) {
+		LOG_ERR("Cannot disable REST mode");
+		return err;
+	}
+
+	/* Write 0x1D in SROM_enable register to initialize the operation */
+	err = reg_write(dev_data, PMW3360_REG_SROM_ENABLE, 0x1D);
+	if (err) {
+		LOG_ERR("Cannot initialize SROM");
+		return err;
+	}
+
+	return err;
+}
+
+static int pmw3360_async_init_fw_load_continue(struct pmw3360_data *dev_data)
+{
+	int err;
+
+	LOG_INF("Uploading optical sensor firmware...");
+
+	/* Write 0x18 to SROM_enable to start SROM download */
+	err = reg_write(dev_data, PMW3360_REG_SROM_ENABLE, 0x18);
+	if (err) {
+		LOG_ERR("Cannot start SROM download");
+		return err;
+	}
+
+	/* Write SROM file into SROM_Load_Burst register.
+	 * Data must start with SROM_Load_Burst address.
+	 */
+	err = burst_write(dev_data, PMW3360_REG_SROM_LOAD_BURST,
+			  pmw3360_firmware_data, pmw3360_firmware_length);
+	if (err) {
+		LOG_ERR("Cannot write firmware to sensor");
+	}
+
+	return err;
+}
+
+static int pmw3360_async_init_fw_load_verify(struct pmw3360_data *dev_data)
+{
+	int err;
+
+	/* Read the SROM_ID register to verify the firmware ID before any
+	 * other register reads or writes
+	 */
+
+	u8_t fw_id;
+	err = reg_read(dev_data, PMW3360_REG_SROM_ID, &fw_id);
+	if (err) {
+		LOG_ERR("Cannot obtain firmware id");
+		return err;
+	}
+
+	LOG_DBG("Optical chip firmware ID: 0x%x", fw_id);
+	if (fw_id != PMW3360_FIRMWARE_ID) {
+		LOG_ERR("Chip is not running from SROM!");
+		return -EIO;
+	}
+
+	u8_t product_id;
+	err = reg_read(dev_data, PMW3360_REG_PRODUCT_ID, &product_id);
+	if (err) {
+		LOG_ERR("Cannot obtain product id");
+		return err;
+	}
+
+	if (product_id != PMW3360_PRODUCT_ID) {
+		LOG_ERR("Invalid product id!");
+		return -EIO;
+	}
+
+	/* Write 0x20 to Config2 register for wireless mouse design.
+	 * This enables entering rest modes.
+	 */
+	err = reg_write(dev_data, PMW3360_REG_CONFIG2, 0x20);
+	if (err) {
+		LOG_ERR("Cannot enable REST modes");
+	}
+
+	return err;
+}
+
+static void irq_handler(struct device *gpiob, struct gpio_callback *cb,
+			u32_t pins)
+{
+	sensor_trigger_handler_t handler;
+
+	k_spinlock_key_t key = k_spin_lock(&pmw3360_data.lock);
+	handler = pmw3360_data.data_ready_handler;
+	k_spin_unlock(&pmw3360_data.lock, key);
+
+	struct sensor_trigger trig = {
+		.type = SENSOR_TRIG_DATA_READY,
+		.chan = SENSOR_CHAN_ALL,
+	};
+	handler(DEVICE_GET(pmw3360), &trig);
+}
+
+static int pmw3360_async_init_power_up(struct pmw3360_data *dev_data)
+{
+	/* Reset sensor */
+
+	spi_cs_ctrl(dev_data, false);
```

Comment:
This is a bit strange, CS should be normally in its inactive state. If it is really necessary to do this call, a comment explaining it could help.

Reply:
I don't think this is necessary as we set the pin in `pmw3360_init_cs` now.

Reply:
fixed